Introduce module tests for GTFS-RT

leonardehrenfried opened the pull request:
### Summary

As discussed in the realtime working group meeting, this introduces module tests for the GTFS-RT realtime code. They are ports of the existing tests but have been converted to use @habrahamsson-skanetrafiken's test environment code. This results in faster test execution and a less confusing debugging experience.

To make this possible, I extracted the bulk of the test environment into a base class which is subclassed by both the Siri and GTFS-RT test to add a few specific methods.

### Unit tests

Well... yes.

### Documentation

Javadoc for each test case.

habrahamsson-skanetrafiken commented:
@leonardehrenfried I think the idea looks good! But in general I would prefer composition in favor of inheritance. Specifically if we ever want to test gtfs and siri together it would get clunky with an abstract base class.

leonardehrenfried commented:
@habrahamsson-skanetrafiken I've changed the inheritance to composition. It's not super-elegant either but less bad than some other options. My intention is for the split between Siri and GTFS test environment to become unnecessary once I've extracted a common timetable snapshot manager.

habrahamsson-skanetrafiken commented on src/ext-test/java/org/opentripplanner/ext/siri/SiriRealtimeTestEnvironment.java:
```diff
@@ -0,0 +1,95 @@
+    return new EntityResolver(testData.getTransitService(), testData.getFeedId());
+  }
+
+  public TripPattern getPatternForTrip(FeedScopedId tripId) {
+    return getPatternForTrip(tripId, RealtimeTestData.SERVICE_DATE);
+  }
+
+  public TripPattern getPatternForTrip(FeedScopedId tripId, LocalDate serviceDate) {
+    var transitService = testData.getTransitService();
+    var trip = transitService.getTripOnServiceDateById(tripId);
+    return transitService.getPatternForTrip(trip.getTrip(), serviceDate);
+  }
+
+  /**
+   * Find the current TripTimes for a trip id on the default serviceDate
+   */
+  public TripTimes getTripTimesForTrip(Trip trip) {
+    return testData.getTripTimesForTrip(trip.getId(), RealtimeTestData.SERVICE_DATE);
+  }
+
+  /**
+   * Find the current TripTimes for a trip id on the default serviceDate
+   */
+  public TripTimes getTripTimesForTrip(String id) {
+    return testData.getTripTimesForTrip(testData.id(id), RealtimeTestData.SERVICE_DATE);
+  }
```
Wouldn't it make sense to have these in the Data-class instead? They don't seem specific to Siri to me?

leonardehrenfried replied:
I've re-combined the two classes: https://github.com/opentripplanner/OpenTripPlanner/pull/5871/commits/dfcba06aba3ed34b4a26734c808ca3294020a874

habrahamsson-skanetrafiken commented on src/test/java/org/opentripplanner/updater/trip/GtfsRealtimeTestEnvironment.java:
```diff
@@ -0,0 +1,42 @@
+    );
+  }
+
+  public TripPattern getPatternForTrip(Trip trip) {
+    return testData.transitModel.getTransitModelIndex().getPatternForTrip().get(trip);
+  }
```
Maybe this should be in the Data-class as well?

leonardehrenfried replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/5871/commits/dfcba06aba3ed34b4a26734c808ca3294020a874

habrahamsson-skanetrafiken commented on src/test/java/org/opentripplanner/updater/trip/TimetableSnapshotSourceTest.java:
```diff
@@ -54,12 +54,18 @@
 
-  private final boolean fullDataset = false;
-  private byte[] cancellation;
+  private final boolean FULL_DATASET = false;
```
Maybe this should be called `NOT_FULL_DATASET` to make it clearer that we are setting the param to false in the call sites?

t2gran replied:
What does "not full" mean? Partial ? I guess the underlying problem is that the input param i a boolean and not an enum.

leonardehrenfried replied:
Thomas is correct. I wanted to turn this boolean into an enum for many years. Do you want me to do that in this PR or another one?

habrahamsson-skanetrafiken commented on src/test/java/org/opentripplanner/updater/trip/TimetableSnapshotSourceTest.java:
```diff
@@ -68,28 +74,17 @@ public void setUp() {
-      fullDataset,
-      List.of(TripUpdate.parseFrom(cancellation)),
+      FULL_DATASET,
+      List.of(CANCELLATION),
```
Much nicer! Why was this encoded and parsed in the original code?

leonardehrenfried replied:
This test is very old. It possible to trace it back to 2012. My guess is that once upon a time, the class under test operated on binary messages, rather than parsed ones. Once the parsing was moved out of it, the test was just quickly patched up rather than properly rewritten.

habrahamsson-skanetrafiken reviewed (changes requested):
If we would want to merge the SiriEnv, the GtfsEnv and the RealtimeTestData into one class, then we could put two separate snapshot sources in the merged environment but make sure only one was initialized. And then throw a runtime exception if you try to call the gtfs methods on an environment where the siri snapshot source was initialized. I think this would work?

But maybe it's nice with the separation that we get now. I don't really have a strong opinion. So you can leave it as it is if you want.

I like the structure that the moduletests add!

Otherwise just some minor comments.

t2gran commented on src/ext-test/java/org/opentripplanner/ext/siri/SiriRealtimeTestEnvironment.java:
```diff
@@ -0,0 +1,95 @@
+  /**
+   * Find the current TripTimes for a trip id on the default serviceDate
+   */
+  public TripTimes getTripTimesForTrip(String id) {
```
```suggestion
  public TripTimes getTripTimesForTrip(String tripId) {
```

leonardehrenfried replied:
That's changed in the latest version.
